### ISSUE TITLE
[BUG] Missing cookie, infinite refresh

### ISSUE THREAD

**Issue by Richard87** [clerk/javascript#2115]
### Preliminary Checks

- [X] I have reviewed the documentation: https://clerk.com/docs

- [X] I have searched for existing issues: https://github.com/clerk/javascript/issues

- [X] I have not already reached out to Clerk support via email or Discord (if you have, no need to open an issue here)

- [X] This issue is not a question, general help request, or anything other than a bug report directly related to Clerk. Please ask questions in our Discord community: https://clerk.com/discord.

### Reproduction / Replay Link

https://pr112.staging.happydogs.no/     https://pr114.staging.happydogs.no/

### Publishable key

pk_test_cGVyZmVjdC1nb3JpbGxhLTk4LmNsZXJrLmFjY291bnRzLmRldiQ

### Description

Bage keeps reloading.

Works in Local env and Prod env. but not in staging env:

```js
    window.__clerk_debug = {
        "frontendApi": "perfect-gorilla-98.clerk.accounts.dev",
        "isSignedIn": false,
        "proxyUrl": "",
        "isInterstitial": true,
        "reason": "cookie-missing",
        "message": "",
        "publishableKey": "pk_test_cGVyZmVjdC1nb3JpbGxhLTk4LmNsZXJrLmFjY291bnRzLmRldiQ",
        "isSatellite": false,
        "domain": ""
    };
```
Our middleware:
```js
import { authMiddleware } from "@clerk/nextjs";

export default authMiddleware({
  publicRoutes: ["/(.*)"],
  audience: "https://happydogs.no",
  // Some comment
});

export const config = {
  matcher: ["/((?!.*\\..*|_next).*)", "/", "/(api|trpc)(.*)"],
};
```

Debug logs:
```
Clerk debug start :: authMiddleware
-- URL debug, {
  "url": "http://localhost:3000/",
  "method": "GET",
  "headers": "{...__clerk_db_jwt=\...\"x-forwarded-for\":\"10.244.6.1,::ffff:10.244.6.28\",\"x-forwarded-host\":\"pr111.staging.happydogs.no\",\"x-forwarded-port\":\"443,80\",\"x-forwarded-proto\":\"https,http\",\"x-forwarded-server\":\"traefik-5b7d576f5f-k5h6r\",\"x-invoke-path\":\"\",\"x-invoke-query\":\"\",\"x-middleware-invoke\":\"1\",\"x-real-ip\":\"10.244.6.1\"}",
  "nextUrl": "http://localhost:3000/",
  "clerkUrl": "https://pr111.staging.happydogs.no/"
}
-- Options debug, {
  "audience": "https://happydogs.no",
  "debug": true,
  "beforeAuth": false,
  "afterAuth": false
}
-- authenticateRequest state is interstitial, {
  "status": "interstitial",
  "reason": "cookie-missing",
  "message": "",
  "frontendApi": "perfect-gorilla-98.clerk.accounts.dev",
  "publishableKey": "pk_test_cGVyZmVjdC1nb3JpbGxhLTk4LmNsZXJrLmFjY291bnRzLmRldiQ",
  "isSatellite": false,
  "domain": "",
  "proxyUrl": "",
  "signInUrl": "",
  "isSignedIn": false,
  "isInterstitial": true,
  "isUnknown": false
}
2023-11-09T20:14:14.064586032Z
```

(No feedback from discord... https://discord.com/channels/856971667393609759/1172250433599852555)

### Environment

```Shell
FROM node:20-alpine AS base

# Install dependencies only when needed
FROM base AS deps
RUN apk add --no-cache libc6-compat
WORKDIR /app

# Install dependencies based on the preferred package manager
COPY package.json package-lock.json*  ./
RUN npm ci


# Rebuild the source code only when needed
FROM base AS builder
WORKDIR /app
ARG NODE_ENV
ARG NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY
ENV NEXT_TELEMETRY_DISABLED 1
RUN echo "Using env: $NODE_ENV"
COPY --from=deps /app/node_modules ./node_modules
COPY . .

RUN npm run build

# If using npm comment out above and use below instead
# RUN npm run build

# Production image, copy all the files and run next
FROM base AS runner
WORKDIR /app
ARG NODE_ENV
# Uncomment the following line in case you want to disable telemetry during runtime.
ENV NEXT_TELEMETRY_DISABLED 1

RUN addgroup --system --gid 1001 nodejs
RUN adduser --system --uid 1001 nextjs

COPY --from=builder /app/public ./public

# Set the correct permission for prerender cache
RUN mkdir .next
RUN chown nextjs:nodejs .next

COPY --from=builder --chown=nextjs:nodejs /app/.next/standalone ./
COPY --from=builder --chown=nextjs:nodejs /app/.next/static ./.next/static

USER nextjs
EXPOSE 3000
ENV PORT 3000
ENV HOSTNAME "0.0.0.0"
CMD ["node", "server.js"]

```
```

**Comment by weisisheng** [comment]
FWIW: Believe I had the same problem, downgrading from Next.js 14.x to 13.5.6 and clerk 4.27.1 to 4.25.7 got me on track again.

**Comment by stx-chris** [comment]
Same here; downgrading to Next.js `13.5.6` stops the infinite loop for us. As of `v14`, e.g. `14.0.2` this issue can be observed on the built app.

**Comment by tjl-jg** [comment]
@weisisheng @stx-chris 

Did the error you experienced appear like the Clerk login components never rendering? Or something else? In testing, we are observing occasional instances of the Clerk login components never rendering, and it being very hard to dislodge/fix the corrupted state. Page refreshes and browser hard refreshes do not seem to be sufficient.

We have not been able to reproduce in a development environment.

**Comment by stx-chris** [comment]
Right, they never render. Instead, an infinite loading loop occurs. I guess I found the culprit:

The **debug** log of the `authMiddleware` states 
```
{ ...
  url: "https://localhost:3000" <--- SSL, not working
  ...
  reason: "cross-origin-referrer"
}
```

in our Next.js `v14.0.2` instance, but

```
{ ...
  url: "http://localhost:3000" <---- non-SSL, working
  ...
  reason: "cross-origin-referrer"
}
```
in our Next.js `v13.5.6` instance. 

For some reason, Next 14 communicates (or Clerk receives). the protocol `https` instead of `http`. Any ideas?

**Comment by stx-chris** [comment]
Our issue seems to be fixed in the latest Next.js canary release `14.0.3-canary.9`.

**Comment by BRKalow** [comment]
@stx-chris Thanks for the info here. Just to confirm, this is in a deployed app using a Clerk production instance running in a Docker container?

**Comment by weisisheng** [comment]
> @weisisheng @stx-chris
> 
> Did the error you experienced appear like the Clerk login components never rendering? Or something else? In testing, we are observing occasional instances of the Clerk login components never rendering, and it being very hard to dislodge/fix the corrupted state. Page refreshes and browser hard refreshes do not seem to be sufficient.
> 
> We have not been able to reproduce in a development environment.


They were not rendering, 

console was repeating "Cookie “__client_uat” has been rejected for invalid domain." and "next.js 13 Cookie jwt does not have a proper “SameSite” attribute value."

After, downgrading, I still get the first error but everything renders and functions in dev.

Thx

**Comment by stx-chris** [comment]
> @stx-chris Thanks for the info here. Just to confirm, this is in a deployed app using a Clerk production instance running in a Docker container?

Not quite, we are still in development phase and therefore deployed a development instance running in a Docker container.

**Comment by ohodegaa** [comment]
We are experiencing the same issue here after upgrading to Next.js 14, running in dev mode locally in a Docker container.

**Comment by tmilewski** [comment]
@weisisheng @tjl-jg @Richard87 @stx-chris 

Cross-posting this from https://github.com/clerk/javascript/issues/1436#issuecomment-1814008308: 

After digging heavily into Next's internals, as well as performing a number of tests with SSL termination and reverse proxying, I'm fairly confident that https://github.com/vercel/next.js/pull/58500, will solve these issues.

Currently, the PR is approved and, once merged, the changes will be available as a canary build of v14.0.3.

In the meantime, you should should be able to mitigate these issues by setting the X-Forwarded-Host request header to the Host header via Varnish, or otherwise.

This was my configuration via [Caddy](https://caddyserver.com/docs/):

```conf
localhost {
  reverse_proxy localhost:3000 {
    header_up X-Forwarded-Host {host}
  }
}
```

In short, if X-Forwarded-Host is already set, Next doesn't overwrite it with the incorrect value.

**Comment by tmilewski** [comment]
@weisisheng @tjl-jg @Richard87 @stx-chris Following up! Please try upgrading to Next v14.0.3, which includes our PR from the other day. This should correct the issues you're experiencing.

Thanks!

**Comment by stx-chris** [comment]
Hi @tmilewski I can confirm that our infinite loop issue has been fixed with this release, thank you very much!

**Comment by tmilewski** [comment]
Thanks all! I'm going to close this. If you have any other issues, feel free to re-open.

**Comment by clerk-cookie** [comment]
This issue has been automatically locked since there has not been any recent activity after it was closed. Please open a new issue for related bugs.
